### ISSUE TITLE
[RNNLM] Fail with large vocab on billion-word LM

### ISSUE THREAD

**Issue by francoishernandez** [kaldi-asr/kaldi#2068]
Hi,
As part of our experiments with Kaldi-RNNLM, we wanted to train a baseline on the One Billion Word LM benchmark dataset.
Trying with a vocab of ~800k it keeps failing at the first 'compute-prob' iteration. On an extract of ~20M words with the same vocab, it fails too. But on the same extract, using Tedlium's 152k vocab, it works.
Is such a big vocab too large for such models?

Thanks in advance!
F

Here are the logs from `compute_prob.0.log`:
```
# rnnlm-get-egs --bos-symbol=793471 --brk-symbol=793473 --eos-symbol=793472 --vocab-size=793474 exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B/text/dev.txt ark:- | rnnlm-compute-prob --use-gpu=yes exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B/0.raw "rnnlm-get-word-embedding exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B/word_feats.txt exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B/feat_embedding.0.mat -|" ark:- 
# Started at Thu Dec  7 00:32:09 CET 2017
#
rnnlm-compute-prob --use-gpu=yes exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B/0.raw 'rnnlm-get-word-embedding exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B/word_feats.txt exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B/feat_embedding.0.mat -|' ark:- 
rnnlm-get-egs --bos-symbol=793471 --brk-symbol=793473 --eos-symbol=793472 --vocab-size=793474 exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B/text/dev.txt ark:- 
LOG (rnnlm-get-egs[5.3.4~1-4ac4]:Process():rnnlm-example.cc:703) Processed 6075 lines of input.
WARNING (rnnlm-compute-prob[5.3.4~1-4ac4]:SelectGpuId():cu-device.cc:183) Not in compute-exclusive mode.  Suggestion: use 'nvidia-smi -c 3' to set compute exclusive mode
LOG (rnnlm-compute-prob[5.3.4~1-4ac4]:SelectGpuIdAuto():cu-device.cc:302) Selecting from 1 GPUs
LOG (rnnlm-compute-prob[5.3.4~1-4ac4]:SelectGpuIdAuto():cu-device.cc:317) cudaSetDevice(0): GeForce GTX 1080 Ti	free:10630M, used:540M, total:11171M, free/total:0.951577
LOG (rnnlm-compute-prob[5.3.4~1-4ac4]:SelectGpuIdAuto():cu-device.cc:366) Trying to select device: 0 (automatically), mem_ratio: 0.951577
LOG (rnnlm-compute-prob[5.3.4~1-4ac4]:SelectGpuIdAuto():cu-device.cc:385) Success selecting device 0 free mem ratio: 0.951577
LOG (rnnlm-compute-prob[5.3.4~1-4ac4]:FinalizeActiveGpu():cu-device.cc:234) The active GPU is [0]: GeForce GTX 1080 Ti	free:10504M, used:666M, total:11171M, free/total:0.940298 version 6.1
rnnlm-get-word-embedding exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B/word_feats.txt exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B/feat_embedding.0.mat - 
ASSERTION_FAILED (rnnlm-compute-prob[5.3.4~1-4ac4]:Resize():cu-sparse-matrix.cc:275) : 'nnz >= 0 && nnz <= num_rows * num_cols' 

[ Stack-Trace: ]

kaldi::MessageLogger::HandleMessage(kaldi::LogMessageEnvelope const&, char const*)
kaldi::MessageLogger::~MessageLogger()
kaldi::KaldiAssertFailure_(char const*, char const*, int, char const*)
kaldi::CuSparseMatrix<float>::Resize(int, int, int, kaldi::MatrixResizeType)
kaldi::CuSparseMatrix<float>::CuSparseMatrix(kaldi::CuArray<int> const&, kaldi::CuVectorBase<float> const&, int, kaldi::MatrixTransposeType)
kaldi::rnnlm::GetRnnlmExampleDerived(kaldi::rnnlm::RnnlmExample const&, bool, kaldi::rnnlm::RnnlmExampleDerived*)
main
__libc_start_main
rnnlm-compute-prob() [0x41b719]

bash: line 1: 30008 Broken pipe             rnnlm-get-egs --bos-symbol=793471 --brk-symbol=793473 --eos-symbol=793472 --vocab-size=793474 exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B/text/dev.txt ark:-
     30009 Aborted                 (core dumped) | rnnlm-compute-prob --use-gpu=yes exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B/0.raw "rnnlm-get-word-embedding exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B/word_feats.txt exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B/feat_embedding.0.mat -|" ark:-
# Accounting: time=26 threads=1
# Ended (code 134) at Thu Dec  7 00:32:35 CET 2017, elapsed time 26 seconds
```

**Comment by misbullah** [comment]
Hi @francoishernandez,
I also get similar issue for Mandarin corpus. 
I use 107653, then I limit the corpus from 300 million line to be 10 million.
Do you get any solution?

Thanks.

**Comment by francoishernandez** [comment]
Hi @misbullah 
Hmm I've got a training running with a ~150k words vocab on the whole lm1b corpus (>200M lines).
I'm wondering whether it could be some special characters / encoding messing with something. I did not observe any compelling proof yet but that may be relevant to your problem with mandarin.

**Comment by misbullah** [comment]
Interesting,
How long it take time for 150k vocabs on the whole lm1b corpus to prepare language model for sampling which part of rnnlm/prepare_rnnlm_dir.sh?
I get stuck on that part when I use whole training data (> 300M lines). Maybe it is memory problem (the RAM on my machine is 64GB).

I did experiment with small vocabs (5k) and limit Mandarin corpus to be 160k lines. I did not get any problem.

Thanks.

**Comment by francoishernandez** [comment]
It took around 30min. I did not monitor memory usage precisely but the machine on which it ran has only 32Gb and there did not seem to have any issue. But it's not the only costly operation in the data prep I presume. It can be a bit long for such amounts of data...
When you say you get stuck, is it failing or is running for a long time? Also, did you have exactly the same error (in compute_prob.0.log)?
Anyways, I'd wait on @danpovey 's comments on this. Not sure of all the implications in the code.

**Comment by misbullah** [comment]
It was running for long time, I was waiting more than 12 hours, then I stopped it.
I got same error but not in compute_prob.0.log file, it is occurred in train.0.1.log (rnnlm-train command).

Thanks.

**Comment by danpovey** [comment]
I think it's an integer overflow issue, where in
'nnz >= 0 && nnz <= num_rows * num_cols'

num_rows * num_cols overflows int32.

Try:

 nnz >= 0 && nnz <= num_rows * static_cast<int64>(num_cols)

and please check in the fix.

**Comment by francoishernandez** [comment]
It appears to solve the assertion problem, but now I have a memory issue due to the size of the objects. Trying to allocate 13Gb on a 1080Ti with only 11... Still, training seems to run fine if we ignore the diagnostics.
Are there some parameters I may tune to reduce memory footprint of the compute-prob jobs?
I checked it in anyways - PR #2077 .

**Comment by danpovey** [comment]
Can you show the details of the error you got?  It may be possible to
reorganize some code to make it scale better.

**Comment by francoishernandez** [comment]
Sure, here are the logs:

```
# rnnlm-get-egs --bos-symbol=793471 --eos-symbol=793472 --brk-symbol=793473 --vocab-size=793474 exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B_toy_fullvocab/text/dev.txt ark:- | rnnlm-compute-prob --use-gpu=yes exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B_toy_fullvocab/0.raw "rnnlm-get-word-embedding exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B_toy_fullvocab/word_feats.txt exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B_toy_fullvocab/feat_embedding.0.mat -|" ark:- 
# Started at Mon Dec 11 11:19:01 CET 2017
#
rnnlm-get-egs --bos-symbol=793471 --eos-symbol=793472 --brk-symbol=793473 --vocab-size=793474 exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B_toy_fullvocab/text/dev.txt ark:- 
rnnlm-compute-prob --use-gpu=yes exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B_toy_fullvocab/0.raw 'rnnlm-get-word-embedding exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B_toy_fullvocab/word_feats.txt exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B_toy_fullvocab/feat_embedding.0.mat -|' ark:- 
LOG (rnnlm-get-egs[5.3.29~1-8e7b5]:Process():rnnlm-example.cc:703) Processed 6075 lines of input.
WARNING (rnnlm-compute-prob[5.3.29~1-8e7b5]:SelectGpuId():cu-device.cc:183) Not in compute-exclusive mode.  Suggestion: use 'nvidia-smi -c 3' to set compute exclusive mode
LOG (rnnlm-compute-prob[5.3.29~1-8e7b5]:SelectGpuIdAuto():cu-device.cc:302) Selecting from 1 GPUs
LOG (rnnlm-compute-prob[5.3.29~1-8e7b5]:SelectGpuIdAuto():cu-device.cc:317) cudaSetDevice(0): GeForce GTX 1080 Ti	free:10807M, used:363M, total:11171M, free/total:0.967421
LOG (rnnlm-compute-prob[5.3.29~1-8e7b5]:SelectGpuIdAuto():cu-device.cc:366) Trying to select device: 0 (automatically), mem_ratio: 0.967421
LOG (rnnlm-compute-prob[5.3.29~1-8e7b5]:SelectGpuIdAuto():cu-device.cc:385) Success selecting device 0 free mem ratio: 0.967421
LOG (rnnlm-compute-prob[5.3.29~1-8e7b5]:FinalizeActiveGpu():cu-device.cc:234) The active GPU is [0]: GeForce GTX 1080 Ti	free:10743M, used:427M, total:11171M, free/total:0.961692 version 6.1
rnnlm-get-word-embedding exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B_toy_fullvocab/word_feats.txt exp/rnnlm_2lstmp_3tdnn_2x256_e1024_1B_toy_fullvocab/feat_embedding.0.mat - 
LOG (rnnlm-compute-prob[5.3.29~1-8e7b5]:PrintMemoryUsage():cu-allocator.cc:127) Memory usage: 3398061716 bytes currently allocated (max: 3398061716); 3351088788 currently in use by user (max: 16351366804); 197/201 calls to Malloc* resulted in CUDA calls.
WARNING (rnnlm-compute-prob[5.3.29~1-8e7b5]:MallocPitchInternal():cu-allocator.cc:97) Allocation of 3173896 x 4096 region failed: freeing some memory and trying again. 
LOG (rnnlm-compute-prob[5.3.29~1-8e7b5]:MallocPitchInternal():cu-allocator.cc:102) To avoid future problems like this, changing memory_factor from 1.5 to 1.1
LOG (rnnlm-compute-prob[5.3.29~1-8e7b5]:PrintMemoryUsage():cu-allocator.cc:127) Memory usage: 3351088788 bytes currently allocated (max: 3398061716); 3351088788 currently in use by user (max: 16351366804); 197/201 calls to Malloc* resulted in CUDA calls.
WARNING (rnnlm-compute-prob[5.3.29~1-8e7b5]:MallocPitchInternal():cu-allocator.cc:97) Allocation of 3173896 x 4096 region failed: freeing some memory and trying again. 
LOG (rnnlm-compute-prob[5.3.29~1-8e7b5]:PrintMemoryUsage():cu-allocator.cc:127) Memory usage: 3351088788 bytes currently allocated (max: 3398061716); 3351088788 currently in use by user (max: 16351366804); 197/201 calls to Malloc* resulted in CUDA calls.
ERROR (rnnlm-compute-prob[5.3.29~1-8e7b5]:MallocPitchInternal():cu-allocator.cc:114) Cannot allocate the requested memory (3173896 x 4096 = 13000278016 bytes)

[ Stack-Trace: ]

kaldi::MessageLogger::HandleMessage(kaldi::LogMessageEnvelope const&, char const*)
kaldi::MessageLogger::~MessageLogger()
kaldi::CuMemoryAllocator::MallocPitchInternal(unsigned long, unsigned long, unsigned long*)
kaldi::CuMemoryAllocator::MallocPitch(unsigned long, unsigned long, unsigned long*)
kaldi::CuMemoryAllocator::MallocPitchLocking(unsigned long, unsigned long, unsigned long*)
kaldi::CuDevice::MallocPitch(unsigned long, unsigned long, unsigned long*)
kaldi::CuMatrix<float>::Resize(int, int, kaldi::MatrixResizeType, kaldi::MatrixStrideType)
kaldi::CuMatrix<float>::CuMatrix(int, int, kaldi::MatrixResizeType, kaldi::MatrixStrideType)

kaldi::rnnlm::ProcessRnnlmOutput(kaldi::rnnlm::RnnlmObjectiveOptions const&, kaldi::rnnlm::RnnlmExample const&, kaldi::rnnlm::RnnlmExampleDerived const&, kaldi::CuMatrixBase<float> const&, kaldi::CuMatrixBase<float> const&, kaldi::CuMatrixBase<float>*, kaldi::CuMatrixBase<float>*, float*, float*, float*, float*)
kaldi::rnnlm::RnnlmCoreComputer::ProcessOutput(kaldi::rnnlm::RnnlmExample const&, kaldi::rnnlm::RnnlmExampleDerived const&, kaldi::CuMatrixBase<float> const&, kaldi::nnet3::NnetComputer*, kaldi::CuMatrixBase<float>*, float*)
kaldi::rnnlm::RnnlmCoreComputer::Compute(kaldi::rnnlm::RnnlmExample const&, kaldi::rnnlm::RnnlmExampleDerived const&, kaldi::CuMatrixBase<float> const&, float*, kaldi::CuMatrixBase<float>*)
main
__libc_start_main
rnnlm-compute-prob() [0x41b719]

LOG (rnnlm-compute-prob[5.3.29~1-8e7b5]:PrintStatsOverall():rnnlm-core-training.cc:118) Overall objf is (-nan + -nan) = -nan over 0 words (weighted) in 0 minibatches; exact = (-nan + -nan) = -nan
LOG (rnnlm-compute-prob[5.3.29~1-8e7b5]:~CachingOptimizingCompiler():nnet-optimize.cc:714) 0.125 seconds taken in nnet3 compilation total (breakdown: 0.0411 compilation, 0.0629 optimization, 0.00343 shortcut expansion, 0.0151 checking, 2.86e-06 computing indexes, 0.00223 misc.)

# Accounting: time=21 threads=1
# Ended (code 255) at Mon Dec 11 11:19:22 CET 2017, elapsed time 21 seconds
```

**Comment by danpovey** [comment]
You seem to be running the "no-sampling" version.  You should be running
the version with sampling.  Even if we fixed the memory problem it would be
super slow.

**Comment by francoishernandez** [comment]
Hmm. There was some sampling when I ran it I think. Maybe there was a problem during preparation, I'll try again. By the way, may the parameter `--unigram-factor` help memory-wise?

**Comment by danpovey** [comment]
Oh, I see-- this is in the part where it's computing the diagnostics.  The
training is probably using sampling.
I'm thinking about this...

**Comment by danpovey** [comment]
@hainan-xv, can you please look into what it would take to, as an option
somewhere in the scripts, have it compute the diagnostics with sampling?
We could maybe have this option turn on automatically - and print a
message- if the number of words is larger than a certain amount (e.g.
100k).
I don't know offhand whether the rnnlm-compute-prob code will just work, or
if we'll need to make some changes; and whether the scripts will just work,
or whether the way it parses the logprobs from the logs would have to be
changed.

And Francois, if you could share with Hainan-- by email maybe-- how your
setup works, it would be helpful too, as he might then be able to replicate
your problem and see if his solution works.  If not, it's OK.

**Comment by vince62s** [comment]
Dan, are you suggesting that at inference there is no importance sampling at all ?

**Comment by danpovey** [comment]
@vince62s, When doing lattice generation, by default we run "unnormalized"-
we use unnormalized logprobs, which doesn't even require sampling-- the
denominator sum is close to 1 due to how we trained it, so this doesn't
make a big difference.

It just happens that the way we set up the training scripts, the part of
the script that gets the diagnostics during training doesn't have the
sampling turned on.  That's because I wanted to see the sampled and
non-sampled objf to get a feeling for how the sampling was working, and
because our initial setups didn't have huge vocabs so it was no problem to
do that.  Hainan is going to look into changing the diagnostic scripts
(compute_prob.*.log, IIRC) to use sampling.
